index.rst

```diff
@@ -13,17 +13,256 @@
+
+``Butler.put``
+""""""""""""""
+
+1. Obtain a valid expanded data ID from the Registry.
+   In the vast majority of cases (i.e. QuantumGraph execution) this done well in advance of the actual ``put`` call.
+
+2. Construct a ``DatasetRef`` by generating an appropriate UUID and using an existing or soon-to-exist RUN name.
+   This will also typically occur well before the rest of the ``put``, as part of QuantumGraph generation.
+
+3. Perform the ``Datastore.put`` operation, writing the file artifacts associated with the dataset and returning records to the Butler.
+   Datastore can be expected to make this operation atomic, either because it is naturally atomic for its storage backing or via writing a temporary and moving it.
+   We do have to accept the possibility of failures leaving partially-written temporary files around.
+   When the Datastore is backed by storage that requires signed URLs, this write is to a staging area and uses non-signed URIs.
+
+4. If the butler has a Registry, either held directly (as in a "full butler" today) or as a client of a butler server, call ``Registry.insertDatasets`` with both the ``DatasetRef`` and the records returned by the Datastore.
+   Database transactions can be used to ensure that all tables in the Registry (including those for datastore records) are updated consistently or left unchanged.
+   If this operation fails, or the butler does not have a Registry (e.g. ``QuantumBackedButler``), the dataset is left in a valid state: it is in the Datastore, but not the Registry.
+   If the Datastore is backed by storage that requires signed URLs, this call also tells the server to transfer the artifacts from the staging area to permanent storage after checking for permissions.
+   This must happen before the database changes are committed.
+
+This has two major advantages over our current ``put`` implementation:
+
+- there is no database transaction over the Datastore write, keeping transactions small and reducing contention for database connections;
+
+- for a client/server butler, there is no alternation between object-store and http operations, just one and then the other, reducing latency (assuming the data ID has indeed been obtained in advance), and increasing the possiblity that the client Datastore can just be a regular ``FileDatastore``.
+  Any database transactions needed can also happen entirely in a single server operation.
+
+This change will make it so Datastore sees conflicting writes before Registry.
+While these are expected to be rare, and thus conflict resolution does not need to be highly optimized, is important that it does not have the potential to corrupt the repository.
+The easiest way to do this is to configure Datastore storage backends to prohibit writes to locations where files already exist; this is the default behavior for most backends of interest, but we currently explicitly permit silent clobbering to work around problems with automatic retries in BPS.
```

Comment:
No. We always allow overwrite now. Allowing for no-overwrite can sometime result in obvious races and as I say above I don't think `os.rename` for local POSIX datastores has the ability to prevent rename to an existing file (short of using chmod/umask on every file in the datastore as we write it to make it read-only -- requiring deletes to chmod it back to read/write).

Reply:
Um, I believe rename is purely a directory operation.  It pays no attention to file permissions.

Reply:
I think the text agrees with @timj's comment for object stores; I think those "obvious races" go away when we require the Datastore write to succeed before we start (let alone commit) the database transaction.

But as Tim noted in another comment, we may indeed have a problem for POSIX.  I haven't looked hard for solutions there.

And the possibility of new and old butler code writing to the same data repository with Datastore/Registry writes in the opposite order would indeed present a challenging transition.

Reply:
This is now better addressed in the text, as per our more recent Slack conversation.